dali/pipeline/util/bounding_box.h

```diff
@@ -20,170 +20,282 @@
-    return *this;
+    CheckBounds(bbox_bounds[0], bbox_bounds[ndim], limits[0], limits[ndim], "left", "right");
+    CheckBounds(bbox_bounds[1], bbox_bounds[ndim+1], limits[1], limits[ndim+1], "top", "bottom");
+
+    // handle depth dimension (if necessary)
+    if (ndim == 3)
+      CheckBounds(bbox_bounds[2], bbox_bounds[ndim+2], limits[2], limits[ndim+2], "front", "back");
+
+    return BoundingBox{bbox_bounds};
   }
 
-  friend void swap(BoundingBox& lhs, BoundingBox& rhs) {
-    using std::swap;
-    swap(lhs.left_, rhs.left_);
-    swap(lhs.right_, rhs.right_);
-    swap(lhs.top_, rhs.top_);
-    swap(lhs.bottom_, rhs.bottom_);
-    swap(lhs.area_, rhs.area_);
+  static BoundingBox FromStartAndShape(RelBounds start_shape, RelBounds limits = {}) {
+    DALI_ENFORCE(start_shape.size() == limits.size() || limits.empty());
+    DALI_ENFORCE(start_shape.size() % 2 == 0);
+    int ndim = start_shape.size() / 2;
+    DALI_ENFORCE(ndim >= 2 && ndim <= 3, make_string("Unexpected number of dimensions: ", ndim));
+
+    auto bbox_bounds = start_shape;
+    for (int i = 0; i < ndim; i++) {
+      bbox_bounds[ndim+i] += bbox_bounds[i];
+    }
+
+    return FromStartAndEnd(bbox_bounds, limits);
   }
 
-  static BoundingBox FromLtrb(const float* data, array<float, kSize> bounds = UniformSquare()) {
-    return FromLtrb(data[0], data[1], data[2], data[3], bounds);
+  static BoundingBox FromLtrb(float l, float t, float r, float b,
+                              RelBounds limits = {}) {
+    return FromStartAndEnd(RelBounds{l, t, r, b}, limits);
   }
 
-  static BoundingBox FromLtrb(
-    float l, float t, float r, float b, array<float, kSize> bounds = UniformSquare()) {
-    CheckBounds(l, bounds[0], bounds[2], "left");
-    CheckBounds(r, bounds[0], bounds[2], "right");
-    CheckBounds(t, bounds[1], bounds[3], "top");
-    CheckBounds(b, bounds[1], bounds[3], "bottom");
+  static BoundingBox FromXywh(float x, float y, float w, float h,
+                              RelBounds limits = {}) {
+    return FromStartAndShape(RelBounds{x, y, w, h}, limits);
+  }
 
-    DALI_ENFORCE(
-      l <= r, "Expected left <= right. Received: " + to_string(l) + " <= " + to_string(r));
-    DALI_ENFORCE(
-      t <= b, "Expected top <= bottom. Received: " + to_string(t) + " <= " + to_string(b));
+  BoundingBox Intersect(const BoundingBox& oth) const {
+    DALI_ENFORCE(ndim() == oth.ndim());
+    RelBounds out;
+    if (!Overlaps(oth)) {
+      return out;
+    }
+    out.reserve(ndim() * 2);
+    for (int i = 0; i < ndim(); i++)
+      out.push_back(std::max(oth.bbox_bounds_[i], bbox_bounds_[i]));
 
-    return {l, t, r, b};
-  }
+    for (int i = 0; i < ndim(); i++)
+      out.push_back(std::min(oth.bbox_bounds_[ndim() + i], bbox_bounds_[ndim() + i]));
 
-  static BoundingBox FromXywh(const float* data, array<float, kSize> bounds = UniformSquare()) {
-    return FromXywh(data[0], data[1], data[2], data[3], bounds);
+    return BoundingBox{out};
```

Comment:
It may not work, if you have a such pair [0,1] and [2,3] then your intersection would be [2,1] which invalid bbox.